designs/2024-config-lookup-from-file/README.md

```diff
@@ -0,0 +1,219 @@
+
+### If I'm linting `./subdir/foo.js` where `./subdir/eslint.config.js` exists, why is `./eslint.config.js` loaded?
+
+Here's the structure in question:
+
+```
+/usr/tmp/
+├── eslint.config.js
+└── subdir/
+    ├── foo.js
+    └── eslint.config.js
+```
+
+Because ignores are handled in `eslint.config.js`, we need to look at `./eslint.config.js` to ensure that `subdir` isn't being ignored.
+
+### What happens if there's a `eslint.config.js` file with an ignore pattern of `subdir` in the parent directory `subdir`?
+
+Here's the structure in question:
+
+```
+/usr/tmp/
+├── eslint.config.js    <-- ignores: ["subdir"]
+└── subdir/
+    ├── foo.js
+    └── eslint.config.js
+```
+
+In this scenario, it matters how ESLint is called:
+
+1. `eslint .` - ESLint first reads `./eslint.config.js`, which says to ignore `subdir`. Because ESLint must recursively traverse `.`, it skips `subdir`. Only `./eslint.config.js` is linted.
+1. `eslint subdir` - ESLint first reads `./subdir/eslint.config.js` and lints `./subdir/foo.js` and `./subdir/eslint.config.js`. The `./eslint.config.js` file is never read because config lookup starts and stops in `subdir` due to the presence of `./subdir/eslint.config.js`.
+1. `eslint subdir/foo.js` - ESLint first reads `./subdir/eslint.config.js` and lints `./subdir/foo.js`.
+
+A more complicated example:
+
+```
+/usr/tmp/
+├── eslint.config.js       <-- ignores: ['**/subsubdir1']
+└── subdir/
+    ├── eslint.config.js   <-- ignores: ['**/subsubdir2'] 
+    ├── subsubdir1/
+    │   └── file.js
+    └── subsubdir2/
+        └── file.js
+```
+
+In this scenario:
+
+1. `eslint .` - ESLint first reads `./eslint.config.js`, which has an `ignores` pattern that doesn't match any of the child directories, so ESLint traverses into each child directory. Once in `subdir`, ESLint reads `./subdir/eslint.config.js`, which says to ignore `subsubdir2`, so ESLint skips traversing into `subsubdir2` but still traverses into `subsubdir1`. ESLint lints `./eslint.config.js`, `./subdir/eslint.config.js`, and `./subdir/subsubdir1/file.js`.
+1. `eslint subdir` - ESLint first reads `./subdir/eslint.config.js` which says to ignore `subsubdir2`, so ESLint skips traversing into `subsubdir2` but still traverses into `subsubdir1`. ESLint lints `./subdir/eslint.config.js`, and `./subdir/subsubdir1/file.js`.
+
+### What happens when I run `eslint *`?
+
+Consider this example:
+
+```
+/usr/tmp/
+└── subdir/
+    ├── eslint.config.js
+    ├── subsubdir1/
+    │   └── file.js
+    └── subsubdir2/
+        └── file.js
+```
+
+When you run `eslint *`, it's the same as if you ran `eslint subdir`, as it matches all immediate children of `.`. If the immediate children include `node_modules` and `.git`, those will still be ignored because they are default ignores in ESLint. All other subdirectories will be traversed but will only be linted if there's an `eslint.config.js` file present, as is the case with `./subdir` in this example.
```

Comment:
Globs match only files, so `eslint *` run in `/user/tmp/` should not lint files in `subdir` because `*` doesn't match them.

Reply:
At least on Windows, `*` matches directories when the shell expansion happens on the command line:

```shell
$ node ./bin/eslint.js * --debug
  eslint:cli CLI args: [ 'bin',             'build', 'CHANGELOG.md',    'conf', 'CONTRIBUTING.md', 'coverage', 'docs',            'eslint.config.js', 'knip.jsonc',      'lib', 'LICENSE',         'Makefile.js', 'messages',    
    'node_modules', 'package.json',    'packages', 'README.md',       'SUPPORT.md', 'templates',       'test.js', 'tests',           'tmp', 'tools',           'versions.json', 'wdio.conf.js',    'webpack.config.js', '--debug' ]
```

Reply:
Ah okay. I didn't notice that the pattern was being expanded by the shell in this case.

Reply:
> At least on Windows, `*` matches directories when the shell expansion happens on the command line

Yes, but when the glob is passed through as-is to ESLint, it matches only files. In both flat config and eslintrc modes, `eslint "*"` lints only files in the current working directory. Another example: `eslint "**/subdir"` doesn't lint content of directories named `subdir`; it would only lint files named `subdir`.

Reply:
The question wasn't about `eslint "*"`, though. :smile:

I really don't think it's worth our time to start diving into edge cases at this point. That's what implementation is for.